python/pyspark/sql/datasource.py

```diff
@@ -0,0 +1,185 @@
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterator, Tuple, Union, TYPE_CHECKING
+
+from pyspark.sql import Row
+from pyspark.sql.types import StructType
+
+if TYPE_CHECKING:
+    from pyspark.sql._typing import OptionalPrimitiveType
+
+
+__all__ = ["DataSource", "DataSourceReader"]
+
+
+class DataSource(ABC):
+    """
+    A base class for data sources.
+
+    This class represents a custom data source that allows for reading from and
+    writing to it. The data source provides methods to create readers and writers
+    for reading and writing data, respectively. At least one of the methods ``reader``
+    or ``writer`` must be implemented by any subclass to make the data source either
+    readable or writable.
+
+    After implementing this interface, you can start to load your data source using
+    ``spark.read.format(...).load()`` and save data using ``df.write.format(...).save()``.
+    """
+
+    def __init__(self, options: Dict[str, "OptionalPrimitiveType"]):
+        """
+        Initializes the data source with user-provided options.
+
+        Parameters
+        ----------
+        options : dict
+            A dictionary representing the options for this data source.
+
+        Notes
+        -----
+        This method should not contain any non-serializable objects.
+        """
+        self.options = options
+
+    @property
+    def name(self) -> str:
+        """
+        Returns a string represents the short name of this data source.
+        """
+        return self.__class__.__name__
+
+    def schema(self) -> Union[StructType, str]:
+        """
+        Returns the schema of the data source.
+
+        It can reference the ``options`` field to infer the data source's schema when
+        users do not explicitly specify it. This method is invoked once when calling
+        ``spark.read.format(...).load()`` to get the schema for a data source read
+        operation. If this method is not implemented, and a user does not provide a
+        schema when reading the data source, an exception will be thrown.
+
+        Returns
+        -------
+        schema : StructType or str
+            The schema of this data source or a DDL string represents the schema
+
+        Examples
+        --------
+        Returns a DDL string:
+
+        >>> def schema(self):
+        ...    return "a INT, b STRING"
+
+        Returns a StructType:
+
+        >>> def schema(self):
+        ...   return StructType().add("a", "int").add("b", "string")
+        """
+        ...
```

Comment:
is it the same as `raise NotImplementedError`?

Reply:
This is equivalent to returning None.
```python
def test():
  ...
v = test()
v is None 
> True
```